src/cmd/metadata.rs

```diff
@@ -14,57 +14,89 @@
 // You should have received a copy of the GNU General Public License
 // along with ink!.  If not, see <http://www.gnu.org/licenses/>.
 
-use std::path::PathBuf;
+use crate::{
+    util,
+    workspace::{ManifestPath, Workspace},
+    Verbosity,
+};
+use anyhow::Result;
 
-use anyhow::{Context, Result};
-use cargo_metadata::MetadataCommand;
+const METADATA_FILE: &str = "metadata.json";
 
 /// Executes build of the smart-contract which produces a wasm binary that is ready for deploying.
 ///
 /// It does so by invoking build by cargo and then post processing the final binary.
-pub(crate) fn execute_generate_metadata(dir: Option<&PathBuf>) -> Result<String> {
+pub(crate) fn execute_generate_metadata(
+    manifest_path: ManifestPath,
+    verbosity: Option<Verbosity>,
+) -> Result<String> {
+    util::assert_channel()?;
     println!("  Generating metadata");
 
-    super::exec_cargo(
-        "run",
-        &[
-            "--package",
-            "abi-gen",
-            "--release",
-            // "--no-default-features", // Breaks builds for MacOS (linker errors), we should investigate this issue asap!
-            "--verbose",
-        ],
-        dir,
-    )?;
+    let (metadata, root_package_id) = crate::util::get_cargo_metadata(&manifest_path)?;
 
-    let cargo_metadata = MetadataCommand::new()
-        .exec()
-        .context("Error invoking `cargo metadata`")?;
-    let mut out_path = cargo_metadata.target_directory;
-    out_path.push("metadata.json");
+    let out_path = metadata.target_directory.join(METADATA_FILE);
+    let out_path_display = format!("{}", out_path.display());
+
+    let target_dir = metadata.target_directory.clone();
+
+    let generate_metadata = move |tmp_manifest_path: &ManifestPath| -> Result<()> {
+        let target_dir_arg = format!("--target-dir={}", target_dir.to_string_lossy());
+        util::invoke_cargo(
+            "run",
+            &[
+                "--package",
+                "abi-gen",
+                &tmp_manifest_path.cargo_arg(),
+                &target_dir_arg,
+                "--release",
+                // "--no-default-features", // Breaks builds for MacOS (linker errors), we should investigate this issue asap!
```

Comment:
I think `--no-default-features` is wrong anyway because this tool performs IO, so for example cannot be `no_std`.